heft-plugins/heft-webpack5-plugin/src/Webpack5Plugin.ts

```diff
@@ -148,131 +164,239 @@ export default class Webpack5Plugin implements IHeftTaskPlugin<IWebpack5PluginOp
+      // Determine how we will run the compiler. When serving, we will run the compiler
+      // via the webpack-dev-server. Otherwise, we will run the compiler directly.
+      if (this._serve) {
+        const defaultDevServerOptions: TWebpackDevServer.Configuration = {
+          host: 'localhost',
+          devMiddleware: {
+            publicPath: '/',
+            stats: {
+              cached: false,
+              cachedAssets: false,
+              colors: heftConfiguration.terminalProvider.supportsColor
+            }
+          },
+          client: {
+            logging: 'info'
+          },
+          port: 8080,
+          onListening: (server: TWebpackDevServer) => {
+            const addressInfo: AddressInfo | string | undefined = server.server?.address() as AddressInfo;
+            if (addressInfo) {
+              const address: string =
+                typeof addressInfo === 'string' ? addressInfo : `${addressInfo.address}:${addressInfo.port}`;
+              taskSession.logger.terminal.writeLine(`Started Webpack Dev Server at https://${address}`);
+            }
+          }
+        };
+
+        // Obtain the devServerOptions from the webpack configuration, and combine with the default options
+        let devServerOptions: TWebpackDevServer.Configuration;
+        if (Array.isArray(webpackConfiguration)) {
+          const filteredDevServerOptions: TWebpackDevServer.Configuration[] = webpackConfiguration
+            .map((configuration) => configuration.devServer)
+            .filter((devServer): devServer is TWebpackDevServer.Configuration => !!devServer);
+          if (filteredDevServerOptions.length > 1) {
+            taskSession.logger.emitWarning(
+              new Error(`Detected multiple webpack devServer configurations, using the first one.`)
+            );
+          }
+          devServerOptions = { ...defaultDevServerOptions, ...filteredDevServerOptions[0] };
+        } else {
+          devServerOptions = { ...defaultDevServerOptions, ...webpackConfiguration.devServer };
+        }
 
-      await new Promise<void>((resolve: () => void, reject: (error: Error) => void) => {
-        firstCompilationDoneCallback = resolve;
-        webpackDevServer.start().catch(reject);
-      });
-    } else {
-      if (process.env[WEBPACK_DEV_SERVER_ENV_VAR_NAME]) {
-        logger.emitWarning(
-          new Error(
-            `The "${WEBPACK_DEV_SERVER_ENV_VAR_NAME}" environment variable is set, ` +
-              'which will cause problems when webpack is not running in serve mode. ' +
-              `(Did a dependency inadvertently load the "${WEBPACK_DEV_SERVER_PACKAGE_NAME}" package?)`
-          )
+        // Add the certificate and key to the devServerOptions
+        const certificateManager: CertificateManager = new CertificateManager();
+        const certificate: ICertificate = await certificateManager.ensureCertificateAsync(
```

Comment:
Whether the user wants a dev certificate should probably be controlled by a config file.

Reply:
I'm going to call this out-of-scope, since this would be new functionality where I'm currently just trying to re-implement watch mode. What could be done though is avoiding setting these values if they've already been set, which can be done in the loadWebpackConfig hook by other plugins.